### DISCUSSION TITLE
Access to mapped memory

### DISCUSSION THREAD

**Discussion by boborjan2** [unicorn-engine/unicorn#2032] [Q&A]
I have two questions regarding mapped memory handling:
1) I see there is a database uc_engine->mapped_blocks that contains all user mapped memory (not mmio), specified by uc_mem_map[_ptr]().  This is seemingly used only for checking for memory overlap when mapping new blocks. For converting physical address to MemoryRegion, however, uc->memory_mapping() is used that performs a much more complicated lookup to get the corresponfing mr.
Why is that? These are all physicial memory blocks, they do not change after mapping(?). What am I missing?
Can I use the mapped_blocks only (or similar uplevel structure) to convert a physical address to a memory region (in case I have no mmio)?

2) Injecting data to physical memory: user is supposed to write guest memory via uc_mem_write(). This finds the memory regions affected and performs the actual write. In cases when it is known that a target area is inside a given mr and the host memory is allocated by the user, may the user write guest memory via simple write to the allocated blob (e.g. from a hook), like a DMA would? 
If not, what else shall be done after such a memory write to make emulator states consistent?

Thanks,
Viktor

**Comment by PhilippTakacs** [comment]
`uc_engine->mapped_blocks` is a hack to have simpler access to the mappings. I would avoid to add code based on them. The code behind  `memory_mapping()` is the same what the emulation will do to find the memory region, so the results are consistent with the emulation. Also this includes some optimization so it might be faster then using `uc->mapped_blocks`.

Writing direct to a pointer which is mapped by `uc_mem_map_ptr` is possible. There are some special cases: When you modify executed code, you also need to clean the corresponding TB cache. When you write to some addresses used by the emulated MMU you need to clear the TLB. When using memory snapshots it gets more complex. But for "normal" memory this is not a problem.
